packages/@aws-cdk/aws-ec2/lib/user-data.ts

```diff
@@ -276,3 +276,258 @@ class CustomUserData extends UserData {
+    return this._body;
+  }
+}
+
+/**
+ * Wrapper for `UserData`.
+ */
+class MultipartUserDataPartWrapper extends MultipartUserDataPart {
+  public constructor(public readonly userData: UserData, opts: MultipartUserDataPartWrapperOptions) {
+    super({
+      contentType: opts.contentType || MultipartUserDataPart.DEFAULT_CONTENT_TYPE,
+      // Force Base64 in case userData will contain UTF-8 characters
+      transferEncoding: 'base64',
+    });
+  }
+
+  public get body(): string {
+    // Wrap rendered user data with Base64 function, in case data contains tokens
+    return Fn.base64(this.userData.render());
+  }
+}
+
+/**
+ * Options for creating {@link MultipartUserData}
+ */
+export interface MultipartUserDataOptions {
+  /**
+   * The string used to separate parts in multipart user data archive (it's like MIME boundary).
+   *
+   * This string should contain [a-zA-Z0-9] characters only, and should not be present in any part, or in text content of archive.
+   */
+  readonly partsSeparator: string;
+}
+
+/**
+ * Mime multipart user data.
+ *
+ * This class represents MIME multipart user data, as described in.
+ * [Specifying Multiple User Data Blocks Using a MIME Multi Part Archive](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/bootstrap_container_instance.html#multi-part_user_data)
+ *
+ */
+export class MultipartUserData extends UserData {
+  private static readonly USE_PART_ERROR = 'MultipartUserData does not support this operation. Please add part using addPart.';
+
+  private parts: IMultipart[] = [];
+
+  private opts: MultipartUserDataOptions;
+
+  constructor(opts: MultipartUserDataOptions) {
+    super();
+
+    this.opts = {
+      ...opts,
+    };
+  }
+  /**
+   * Adds existing `UserData`. Modification to `UserData` are reflected in subsequent renders of the part.
+   *
+   * For more information about content types see `MultipartUserDataPartOptionsWithBody`
+   */
+  public addUserDataPart(userData: UserData, opts?: MultipartUserDataPartWrapperOptions): this {
+    this.parts.push(MultipartUserDataPart.fromUserData(userData, opts));
+
+    return this;
+  }
+
+  /**
+   * Adds the 'raw' part using provided options.
+   */
+  public addPart(opts: MultipartUserDataPartOptionsWithBody): this {
+    this.parts.push(MultipartUserDataPart.fromRawBody(opts));
+
+    return this;
+  }
+
+  public render(): string {
+    const boundary = this.opts.partsSeparator;
+
+    // Now build final MIME archive - there are few changes from MIME message which are accepted by cloud-init:
+    // - MIME RFC uses CRLF to separate lines - cloud-init is fine with LF \n only
+    var resultArchive = `Content-Type: multipart/mixed; boundary="${boundary}"\n`;
+    resultArchive = resultArchive + 'MIME-Version: 1.0\n';
+
+    // Add parts - each part starts with boundary
+    this.parts.forEach(part => {
+      resultArchive = resultArchive + '\n--' + boundary + '\n' + 'Content-Type: ' + part.contentType + '\n';
```

Comment:
Doesn't it make more sense for a body part to render its own headers?

Then you wouldn't have to make them public getters either, and it's trivial for new body parts to add their own headers.

I like narrow interfaces :).